esmvaltool/preprocessor/_multimodel.py

```diff
@@ -13,42 +25,314 @@
+        oidx = [t_x.index(int((s / 365.) * 12.)) for s in time_redone]
+        # reshape data to include all possible times
+        ndat = np.ma.resize(cube.data, fine_shape)
+        # build the time mask
+        c_ones = np.ones(fine_shape, bool)
+        c_ones[oidx] = False
+        ndat.mask |= c_ones
+
+        # stitch new datas
+        datas.append(ndat)
+
+    # return datas
+    return datas, t_x0
+
+
+def _apply_overlap(cube, tx1, tx2):
+    """
+    Slice cubes
+
+    This operation is memory-intensive;
+    do it ONLY IF you need to do it;
+    otherwise use span == constant.
+    """
+    logger.debug("Bounds: %s and %s", str(tx1), str(tx2))
+    # slice cube on time
+    with iris.FUTURE.context(cell_datetime_objects=True):
+        cube = _slice_cube(cube, tx1, tx2)
+
+    return cube
+
+
+def _assemble_overlap_data(selection, ovlp, stat_type, fname):
+    """Get statistical data in iris cubes for OVERLAP"""
+    start, stop = ovlp
+    utype = str(selection[0].coord('time').units)
+    start_dtime, stop_dtime = [_to_datetime(t, utype) for t in (start, stop)]
+    stats_dats = np.ma.zeros(_slice_cube2(selection[0], start, stop).shape)
+
+    for i in range(stats_dats.shape[0]):
+        time_data = [_slice_cube2(cube, start, stop)[i] for cube in selection]
+        stats_dats[i] = _compute_statistic(time_data, stat_type)
+    stats_cube = _put_in_cube(
+        _apply_overlap(selection[0], start_dtime, stop_dtime),
+        stats_dats,
+        stat_type,
+        fname,
+        t_axis=None)
+    return stats_cube
+
+
+def _assemble_full_data(selection, stat_type, fname):
+    """Get statistical data in iris cubes for FULL"""
+    slices, time_axis = _full_time(selection)
+    stats_dats = np.ma.zeros(slices[0].shape)
+
+    for i in range(slices[0].shape[0]):
+        time_data = [data[i] for data in slices]
+        stats_dats[i] = _compute_statistic(time_data, stat_type)
+    stats_cube = _put_in_cube(selection[0], stats_dats, stat_type, fname,
+                              time_axis)
+    return stats_cube
+
+
+def _update_fname(curr_filename, ovlp_interval, unit_type):
+    """Update netCDF file names based on time properties"""
+    # BACKEND_MultiModel[stats]_[field]_[var]_[start_year]-[end_year].nc
+    froot = "_".join(curr_filename.split('_')[0:-1])
+    start, stop = ovlp_interval
+    yr_1 = str(_to_datetime(start, unit_type).year)
+    yr_2 = str(_to_datetime(stop - 1, unit_type).year)  # -1 for correct leap
+    new_file = "{}_{}-{}.nc".format(froot, yr_1, yr_2)
+    return new_file
+
+
+def multi_model_statistics(cubes, span, filenames, exclude, statistics):
     """Compute multi-model mean and median."""
-    logger.debug('Multi model statistics: excluding files: %s', str(exclude))
+    logger.debug('Multi model statistics: excluding files: %s', exclude)
+
+    logger.debug('Multimodel statistics: computing: %s', statistics)
+    iris.util.unify_time_units(cubes)
```

Comment:
Maybe you could apply `iris.util.unify_time_units(cubes)` only to `selection`? That would be more efficient because then you do not need to convert the time units of excluded cubes.

Reply:
good point! done